### PR TITLE
gh-101714: Isolate  `_curses` extension module

### PR DESCRIPTION
**Pull request by chgnrdv**
#101714

* converted `PyCursesWindowObject` to heap type
* added module state containing previously global `PyCursesWindow_Type`, `PyCursesError` and `screen_encoding`; added `get__curses_state` and `find__curses_state_by_type` functions; changed functions and methods to access members of state or take state as a parameter
* got rid of `ModDict` global variable, made functions/methods access it through `PyModule_GetDict` call; changed `SetDictInt` to accept dict as parameter
* converted module to multi-phase init
* changed signatures of functions used by C API in order to make exception stored in module state accessible by API users; reworked macros in Include/py_curses.h to reflect these changes
* added `clinic_state()` macro for access to `PyCursesWindow_Type` in AC-generated code

### CODE REVIEW

**Comment by chgnrdv** [comment]
@erlend-aasland @kumaraditya303
There are still some things I have doubts about, e.g. how I worked around necessity to expose `PyCursesError` to C API users, e. g. `_curses_panel` module. I see that it could be done in more pretty way through capsule import, in the same manner as between `_elementtree` and `pyexpat`, but I'm not sure about it being non-breaking change.

**Comment by chgnrdv** [comment]
@erlend-aasland @kumaraditya303 for review.

**Comment by erlend-aasland** [comment]
I'm sorry, but I don't have the bandwidth to review this in the near future.

**Comment by chgnrdv** [comment]
@erlend-aasland, no problem, these aren't first priority changes anyway :)

File: Modules/_cursesmodule.c [hunk 1]
```diff
@@ -180,24 +176,71 @@ static int initialised = FALSE;
...
+    _curses_state *state = get_curses_state(mod);
+    Py_VISIT(state->PyCursesWindow_Type);
+    Py_VISIT(state->PyCursesError);
+    return 0;
+}
+
+static void
+_curses_free(void *mod)
+{
+    _curses_clear((PyObject *)mod);
```
Comment (kumaraditya303) [inline Modules/_cursesmodule.c]:
```suggestion
    (void)_curses_clear((PyObject *)mod);
```

File: Modules/_cursesmodule.c [hunk 2]
```diff
@@ -163,13 +163,9 @@ typedef chtype attr_t;           /* No attr_t type is available */
...
 [clinic start generated code]*/
-/*[clinic end generated code: output=da39a3ee5e6b4b0d input=43265c372c2887d6]*/
-
-/* Definition of exception curses.error */
-
-static PyObject *PyCursesError;
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=83369be6e20ef0da]*/
 
 /* Tells whether setupterm() has been called to initialise terminfo.  */
 static int initialised_setupterm = FALSE;
```
Comment (kumaraditya303) [inline Modules/_cursesmodule.c]: These would also have to be moved to module state?

Comment (chgnrdv) [reply Modules/_cursesmodule.c]: Done. Not sure if I did it the best way though.

Comment (chgnrdv) [reply Modules/_cursesmodule.c]: Sure I didn't. At least exposing parts of `_curses` module state to C API in this way doesn't look safe for me.

**Comment by chgnrdv** [comment]
@erlend-aasland, no problem, I was ready for this as I read the [conversation](https://github.com/python/cpython/issues/103092) under original issue and knew that the need to isolate this module is debatable :)
> On the bright side: hopefully, you learnt a lot about the C API, extension modules and how to isolate them!

Sure! Thank you for opportunity :)
I'll see if I can help with one of remaining non-isolated modules.

**Pull request closed without merging**